{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Length of the DNA is', 2872915)\n",
      "['StaphylococcusAureus.fasta', 2872915, 81, 16, 19.491853096329674, True, '2015-12-11 15:34:11.755882']\n",
      "['StaphylococcusAureus.fasta', 2872915, 82, 15, 18.07681559705083, True, '2015-12-11 15:34:31.484442']\n",
      "['StaphylococcusAureus.fasta', 2872915, 83, 15, 18.07681559705083, True, '2015-12-11 15:34:49.198156']\n",
      "['StaphylococcusAureus.fasta', 2872915, 84, 15, 17.491853096329674, True, '2015-12-11 15:35:08.532166']\n",
      "['StaphylococcusAureus.fasta', 2872915, 85, 15, 17.491853096329674, True, '2015-12-11 15:35:32.213978']\n",
      "['StaphylococcusAureus.fasta', 2872915, 86, 15, 17.491853096329674, True, '2015-12-11 15:35:51.365820']\n",
      "['StaphylococcusAureus.fasta', 2872915, 87, 15, 17.491853096329674, True, '2015-12-11 15:36:10.373007']\n",
      "['StaphylococcusAureus.fasta', 2872915, 88, 15, 17.491853096329674, True, '2015-12-11 15:36:30.931749']\n",
      "['StaphylococcusAureus.fasta', 2872915, 89, 14, 15.906890595608518, True, '2015-12-11 15:36:51.746437']\n",
      "['StaphylococcusAureus.fasta', 2872915, 90, 13, 14.906890595608518, True, '2015-12-11 15:37:12.590629']\n",
      "['StaphylococcusAureus.fasta', 2872915, 91, 10, 11.906890595608518, True, '2015-12-11 15:37:32.418131']\n",
      "['StaphylococcusAureus.fasta', 2872915, 92, 10, 11.906890595608518, True, '2015-12-11 15:37:51.278354']\n",
      "['StaphylococcusAureus.fasta', 2872915, 93, 9, 10.906890595608518, True, '2015-12-11 15:38:11.015804']\n",
      "['StaphylococcusAureus.fasta', 2872915, 94, 9, 10.906890595608518, True, '2015-12-11 15:38:36.537406']\n",
      "['StaphylococcusAureus.fasta', 2872915, 95, 8, 9.906890595608518, True, '2015-12-11 15:38:54.396971']\n",
      "['StaphylococcusAureus.fasta', 2872915, 96, 7, 8.906890595608518, True, '2015-12-11 15:39:14.549676']\n",
      "['StaphylococcusAureus.fasta', 2872915, 97, 7, 8.906890595608518, True, '2015-12-11 15:39:35.638947']\n",
      "['StaphylococcusAureus.fasta', 2872915, 98, 7, 8.906890595608518, True, '2015-12-11 15:39:55.742050']\n",
      "['StaphylococcusAureus.fasta', 2872915, 99, 7, 8.906890595608518, True, '2015-12-11 15:40:17.675285']\n",
      "['StaphylococcusAureus.fasta', 2872915, 100, 7, 8.906890595608518, True, '2015-12-11 15:40:37.187516']"
     ]
    }
   ],
   "source": [
    "# %load Genome.py\n",
    "from Bio import SeqIO\n",
    "import sys, numpy, math, datetime\n",
    "import CommonFunctions\n",
    "from CommonFunctions import WriteArrayinFile\n",
    "from operator import itemgetter\n",
    "from collections import Counter\n",
    "\n",
    "class Genome:\n",
    "    DNAList= []\n",
    "    DNA_current = []\n",
    "    ReadLength_Considered = 100\n",
    "    Filename    = \"\"\n",
    "    SideLengths = 1\n",
    "    Gap         = 1\n",
    "    \"\"\"\n",
    "        Basic Initialization\n",
    "    \"\"\"\n",
    "    def __init__(self, Filename, Gap):\n",
    "        self.Gap = Gap\n",
    "        self.Filename = Filename\n",
    "        Handle = open(Filename)\n",
    "        for seq_record in SeqIO.parse(Handle, \"fasta\"):\n",
    "            self.DNAList.append(str(seq_record.seq))\n",
    "        print( \"Length of the DNA is\", len(self.DNAList[0]))\n",
    "    \n",
    "    \n",
    "    def getUncertainty(self):\n",
    "        #Stores final data\n",
    "        Summary = []\n",
    "        length = 80\n",
    "        while True:\n",
    "            length += self.Gap\n",
    "            Answer = self.RepeatsofLengthL(length)\n",
    "            print( Answer)\n",
    "            Summary += [ Answer ]\n",
    "            if Answer[-3] < 1:\n",
    "                break\n",
    "        CommonFunctions.WriteArrayinFile(Summary, \"New_Summary_\"+self.Filename[:-6]+\".csv\")\n",
    "#   \n",
    "    \n",
    "    \"\"\"\n",
    "        Implements log(n!)\n",
    "    \"\"\"\n",
    "    def factlog(self,n):\n",
    "        if n<20:\n",
    "            return(math.log(math.factorial(n), 2))\n",
    "        else:\n",
    "            return( n*(math.log(n,2) - 1.44269) + 0.5*math.log(2*math.pi*n)) #math.log(math.e,2) = 1.4426950408889634\n",
    "    \n",
    "    def RepeatsofLengthL(self, length = 100):\n",
    "        self.ReadLength_Considered = length\n",
    "        self.DNA_current = self.DNAList[0]\n",
    "        #print(\"DNA Length\", len(self.DNA_current))\n",
    "        Reads = []\n",
    "        #print(\"Read Length\", length)\n",
    "        for position in range(self.SideLengths, len(self.DNA_current) - self.ReadLength_Considered - self.SideLengths ):\n",
    "            #if position % 1000000 == 0:\n",
    "            #    print(\"In position\", position, \"DNA left\", len(self.DNA_current) - position )           \n",
    "            Reads +=[   \n",
    "                        [ \n",
    "                         self.DNA_current[position - self.SideLengths:                             position] , #Left Neighbhour\n",
    "                         self.DNA_current[position                   :position + self.ReadLength_Considered] ,\n",
    "                         self.DNA_current[position + self.ReadLength_Considered : position + self.ReadLength_Considered + self.SideLengths], #Right Neigbhour\n",
    "                         position\n",
    "                        ]\n",
    "                    ]\n",
    "        #print(\"Sorting Read Array\")\n",
    "        # Sort w.r.t to the reads. (aggrerate them)\n",
    "        Reads = sorted(Reads,key=itemgetter(1) )\n",
    "        CountInfo = []\n",
    "        ReadInfo = []\n",
    "        \n",
    "        Titles = [ \"Gene\", \"Gene Length\", \"Read Length\", \"Read\", \"Neighbhours\", \"Repeat\"]\n",
    "        Is_less_than_critical_length = False\n",
    "        Reason = \"No reason\"\n",
    "        Position_of_repeat_less_than_2 = []\n",
    "        \n",
    "        #print(\"Counting Repeats\")\n",
    "        CurrentString = Reads[0][1]\n",
    "        Repeat = 0\n",
    "        LeftNeighbhors = []\n",
    "        RightNeighbors = []\n",
    "        RepeatPositions = []\n",
    "        Repeat = 1\n",
    "\n",
    "        for read in Reads[1:]:\n",
    "            if read[1] == CurrentString:\n",
    "                Repeat += 1\n",
    "                LeftNeighbhors  +=   read[0]\n",
    "                RightNeighbors  +=   read[2]\n",
    "                RepeatPositions += [ read[3] ]\n",
    "            else:\n",
    "                \"\"\"\n",
    "                    If the read repeats more than one time and if the right neighbours are unique, \n",
    "                    then the node represented by this read will *not* get condensed and this node will add to the uncertainty\n",
    "                \"\"\"\n",
    "                if Repeat > 1 and len( set(RightNeighbors) ) != 1:\n",
    " #                   print(RepeatPositions)\n",
    "\n",
    "                    #Checking for l_critical\n",
    "                    if Repeat > 2:\n",
    "                        # If repeat >=3, then the given read length is definitely less than L_critical.\n",
    "                        Is_less_than_critical_length = True\n",
    "                        Reason = \"Triple Repeats\"\n",
    "                    else:\n",
    "                        # We accumulate all the reads with length = 2 and check if they interleave.\n",
    "                        Position_of_repeat_less_than_2 += [ RepeatPositions + read[1] ]\n",
    "                    \n",
    "                    # We count the number of times each neighbhour is repeats\n",
    "                    \"\"\"\n",
    "                        Example: If the repeat is \n",
    "                        AACA G\n",
    "                        AACA G\n",
    "                        AACA T\n",
    "                        AACA T\n",
    "                        AACA C\n",
    "                        AACA C\n",
    "                        AACA C\n",
    "                        AACA C\n",
    "                        AACA A\n",
    "                        AACA A\n",
    "                        AACA A\n",
    "                        \n",
    "                        Then the node [AACA] has 2 outgoing edges to ACA G, \n",
    "                        2 outgoing edges to ACA T, 4 outgoing edges to ACA C,\n",
    "                        3 outgoing edges to ACA A.\n",
    "                        \n",
    "                        Note, if AACA is always present as G(AACA), then the node GAAC\n",
    "                        will have all its outgoing edges to AACA and that wont be counted here because\n",
    "                        unique right neigbhour = 1\n",
    "                        \n",
    "                    \"\"\"\n",
    "                    Count_stats = Counter(RightNeighbors).values()\n",
    "                    \n",
    "                    Uncertainty = self.factlog(sum(Count_stats))\n",
    "                    for i in Count_stats:\n",
    "                        Uncertainty -= self.factlog(i)\n",
    "\n",
    "                    CountInfo += [ Uncertainty ]\n",
    "                \n",
    "                \"\"\"\n",
    "                    Re-initialization.\n",
    "                \"\"\"\n",
    "                LeftNeighbhors = []\n",
    "                RightNeighbors = []\n",
    "                RepeatPositions = []\n",
    "                Repeat = 1\n",
    "            \n",
    "                CurrentString = read[1]\n",
    "        \n",
    "        if sorted(Position_of_repeat_less_than_2, key = itemgetter(0)) != sorted(Position_of_repeat_less_than_2, key = itemgetter(1)):\n",
    "           # print( Position_of_repeat_less_than_2, sorted(Position_of_repeat_less_than_2))\n",
    "            Reason = \"Interleaved Repeats\"\n",
    "            Is_less_than_critical_length = True\n",
    "\n",
    "        Summary = [ self.Filename, len(self.DNA_current), self.ReadLength_Considered, len(CountInfo), sum(CountInfo), Is_less_than_critical_length, str(datetime.datetime.now()) ]\n",
    "        if False:\n",
    "            print(\"Length of the DNA is\", len(self.DNA_current))\n",
    "            print(\"Number of reads repeating of length\", self.ReadLength_Considered,\" is\", len(CountInfo))\n",
    "            print(\"Time\", datetime.datetime.now())\n",
    "       # print(\"Uncertainity\", sum(CountInfo), \"Is less than critical length\", Is_less_than_critical_length, \"reason\", Reason)\n",
    "        return(Summary)\n",
    "\n",
    "filename = \"RhodobacterSphaeroides.fasta\"\n",
    "filename = \"Buchnera_aphidicola.fasta\"\n",
    "filename = \"StaphylococcusAureus.fasta\"\n",
    "Gene = Genome(filename, 1)\n",
    "Gene.getUncertainty()\n",
    "\n",
    "\n",
    "#RhodobacterSphaeroides = Genome(\"RhodobacterSphaeroides.fasta\", 3)\n",
    "#RhodobacterSphaeroides.getUncertainty()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "a = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "a = [1,10,7,3,9]\n",
    "sorted(a)\n",
    "print(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
