{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Length of the DNA is', 640681)\n",
      "('Answer', [2, 986804.808506179, 986760.9108821239, True, '2015-12-13 19:24:38.413349'])\n",
      "('Answer', [3, 1077063.8233174058, 1076883.5997124303, True, '2015-12-13 19:24:42.947139'])\n",
      "('Answer', [4, 1111663.9564703598, 1110921.0618202821, True, '2015-12-13 19:24:47.713905'])\n",
      "('Answer', [5, 1120076.9283789415, 1116987.3610080895, True, '2015-12-13 19:24:52.394369'])\n",
      "('Answer', [6, 1096172.78199402, 1083325.5945727662, True, '2015-12-13 19:24:58.394814'])\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "can only concatenate list (not \"str\") to list",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-17-04c7a92c899a>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m    238\u001b[0m \u001b[1;31m#filename = \"StaphylococcusAureus.fasta\"\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    239\u001b[0m \u001b[0mGene\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mGenome\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilename\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;36m1\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 240\u001b[1;33m \u001b[0mGene\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mgetUncertainty\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    241\u001b[0m \u001b[1;31m#Gene.getReadLengthGraph()\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    242\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-17-04c7a92c899a>\u001b[0m in \u001b[0;36mgetUncertainty\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m    111\u001b[0m         \u001b[1;32mwhile\u001b[0m \u001b[0mTrue\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    112\u001b[0m             \u001b[0mlength\u001b[0m \u001b[1;33m+=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mGap\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 113\u001b[1;33m             \u001b[0mAnswer\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mRepeatsofLengthL\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlength\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    114\u001b[0m             \u001b[1;32mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"Answer\"\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mAnswer\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    115\u001b[0m             \u001b[0mSummary\u001b[0m \u001b[1;33m+=\u001b[0m \u001b[1;33m[\u001b[0m \u001b[0mAnswer\u001b[0m \u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-17-04c7a92c899a>\u001b[0m in \u001b[0;36mRepeatsofLengthL\u001b[1;34m(self, length)\u001b[0m\n\u001b[0;32m    173\u001b[0m                     \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    174\u001b[0m                         \u001b[1;31m# We accumulate all the reads with length = 2 and check if they interleave.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 175\u001b[1;33m                         \u001b[0mPosition_of_repeat_less_than_2\u001b[0m \u001b[1;33m+=\u001b[0m \u001b[1;33m[\u001b[0m \u001b[0mRepeatPositions\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mread\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    176\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    177\u001b[0m                     \u001b[1;31m# We count the number of times each neighbhour is repeats\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mTypeError\u001b[0m: can only concatenate list (not \"str\") to list"
     ]
    }
   ],
   "source": [
    "# %load Genome.py\n",
    "# %load Genome.py\n",
    "from Bio import SeqIO\n",
    "import sys, numpy, math, datetime\n",
    "import CommonFunctions\n",
    "from CommonFunctions import WriteArrayinFile\n",
    "from operator import itemgetter\n",
    "from collections import Counter\n",
    "from doctest import Example\n",
    "\n",
    "class Genome:\n",
    "    DNAList= []\n",
    "    DNA_current = []\n",
    "    ReadLength_Considered = 100\n",
    "    Filename    = \"\"\n",
    "    SideLengths = 1\n",
    "    Gap         = 1\n",
    "    \"\"\"\n",
    "        Basic Initialization\n",
    "    \"\"\"\n",
    "    def __init__(self, Filename, Gap):\n",
    "        self.Gap = Gap\n",
    "        self.Filename = Filename\n",
    "        Handle = open(Filename)\n",
    "        for seq_record in SeqIO.parse(Handle, \"fasta\"):\n",
    "            self.DNAList.append(str(seq_record.seq))\n",
    "        print( \"Length of the DNA is\", len(self.DNAList[0]))\n",
    "    \n",
    "    \"\"\"\n",
    "        Implements log(n!)\n",
    "    \"\"\"\n",
    "    def factlog(self,n):\n",
    "        if n<20:\n",
    "            return(math.log(math.factorial(n), 2))\n",
    "        else:\n",
    "            return( n*(math.log(n,2) - 1.44269) + 0.5*math.log(2*math.pi*n)) #math.log(math.e,2) = 1.4426950408889634\n",
    "    \n",
    "    def getReadLengthGraph(self):\n",
    "     #Stores final data\n",
    "        Summary = []\n",
    "        for length in range(1,400):\n",
    "            length += self.Gap\n",
    "            Answer = self.NumberofRepeatsofLengthL(length) \n",
    "            if False == Answer[3]:\n",
    "                break  \n",
    "            print(\"Length\", length, \"Answer\", Answer)\n",
    "            Summary += [ Answer ]\n",
    "        CommonFunctions.WriteArrayinFile(Summary, \"Read_lengths\"+self.Filename[:-6]+\".csv\")\n",
    "           \n",
    "    def NumberofRepeatsofLengthL(self, length):\n",
    "        self.DNA_current = self.DNAList[0]\n",
    "        #print(\"DNA Length\", len(self.DNA_current))\n",
    "        Reads = []\n",
    "        for position in range(self.SideLengths, len(self.DNA_current) - self.ReadLength_Considered - self.SideLengths ):\n",
    "            Reads +=[   \n",
    "                        [ \n",
    "                         self.DNA_current[position - 1: position] , #Left Neighbhour\n",
    "                         self.DNA_current[position: position + length] ,\n",
    "                         self.DNA_current[position + length : position + length + 1],#Right Neigbhour\n",
    "                         position\n",
    "                        ]\n",
    "                    ]\n",
    "        #print(\"Sorting Read Array\")\n",
    "        # Sort w.r.t to the reads. (aggrerate them)\n",
    "        Reads = sorted(Reads,key=itemgetter(1) )\n",
    "        No_of_maximal_reads = 0\n",
    "        No_of_occurences_of_maximal_reads = 0\n",
    "        #print(\"Counting Repeats\")\n",
    "        CurrentString = Reads[0][1]\n",
    "        Neighbhors = []\n",
    "        Repeat = 1\n",
    "        Continue = False\n",
    "        Example = [\"none\"]\n",
    "        Positions = []\n",
    "        for read in Reads[1:]:\n",
    "            if read[1] == CurrentString:\n",
    "                Repeat += 1\n",
    "                Neighbhors += [ (read[0], read[2]) ]\n",
    "                Positions += [ read[3] ]\n",
    "            else:\n",
    "                \"\"\"\n",
    "                   Count only when both right and left neigbhors are not the same\n",
    "                \"\"\"\n",
    "                if Repeat > 1:\n",
    "                    Continue = True\n",
    "                    Example = [ CurrentString]\n",
    "                    Is_Maximal = False\n",
    "                    Neighbhors = list( set(Neighbhors) )\n",
    "                    for i in range(1,len(Neighbhors)):\n",
    "                        for j in range(i):\n",
    "                            if( Neighbhors[i][0] != Neighbhors[j][0] and Neighbhors[i][1] != Neighbhors[j][1] ): #Implies that the given read is maximal\n",
    "                                Is_Maximal = True\n",
    "                                break\n",
    "                        else:\n",
    "                            continue\n",
    "                        break\n",
    "                    if Is_Maximal:\n",
    "                        No_of_maximal_reads += 1\n",
    "                        No_of_occurences_of_maximal_reads += Repeat\n",
    "                            \n",
    "                Repeat = 1        \n",
    "                CurrentString = read[1]\n",
    "                Neighbhors = [ (read[0], read[2]) ]\n",
    "                Positions = [ read[3] ]\n",
    "        return( [ length, No_of_maximal_reads, No_of_occurences_of_maximal_reads, Continue,Example ])\n",
    "\n",
    "    def getUncertainty(self):\n",
    "        #Stores final data\n",
    "        Summary = []\n",
    "        length = 1\n",
    "        while True:\n",
    "            length += self.Gap\n",
    "            Answer = self.RepeatsofLengthL(length)  \n",
    "            print(\"Answer\", Answer)\n",
    "            Summary += [ Answer ]\n",
    "            if Answer[-4] < 1:\n",
    "                break\n",
    "        CommonFunctions.WriteArrayinFile(Summary, \"U_Summary_\"+self.Filename[:-6]+\".csv\")\n",
    "   \n",
    "    def RepeatsofLengthL(self, length = 100):\n",
    "        self.ReadLength_Considered = length\n",
    "        self.DNA_current = self.DNAList[0]\n",
    "        #print(\"DNA Length\", len(self.DNA_current))\n",
    "        Reads = []\n",
    "        #print(\"Read Length\", length)\n",
    "        for position in range(self.SideLengths, len(self.DNA_current) - self.ReadLength_Considered - self.SideLengths ):\n",
    "         #   if position % 1000000 == 0:\n",
    "         #       print(\"In position\", position, \"DNA left\", len(self.DNA_current) - position )           \n",
    "            Reads +=[   \n",
    "                        [ \n",
    "                         self.DNA_current[position - self.SideLengths:                             position] , #Left Neighbhour\n",
    "                         self.DNA_current[position                   :position + self.ReadLength_Considered] ,\n",
    "                         self.DNA_current[position + self.ReadLength_Considered : position + self.ReadLength_Considered + self.SideLengths], #Right Neigbhour\n",
    "                         position\n",
    "                        ]\n",
    "                    ]\n",
    "        #print(\"Sorting Read Array\")\n",
    "        # Sort w.r.t to the reads. (aggrerate them)\n",
    "        Reads = sorted(Reads,key=itemgetter(1) )\n",
    "        UpperboundUncertainty = []\n",
    "        UncertaintyGap = []\n",
    "        \n",
    "        Is_less_than_critical_length = False\n",
    "        Reason = \"No reason\"\n",
    "        Position_of_repeat_less_than_2 = []\n",
    "        \n",
    "        #print(\"Counting Repeats\")\n",
    "        CurrentString = Reads[0][1]\n",
    "        Repeat = 0\n",
    "        LeftNeighbhors = []\n",
    "        RightNeighbors = []\n",
    "        RepeatPositions = []\n",
    "        Repeat = 1\n",
    "\n",
    "        for read in Reads[1:]:\n",
    "            if read[1] == CurrentString:\n",
    "                Repeat += 1\n",
    "                LeftNeighbhors  +=   read[0]\n",
    "                RightNeighbors  +=   read[2]\n",
    "                RepeatPositions += [ read[3] ]\n",
    "            else:\n",
    "                \"\"\"\n",
    "                    If the read repeats more than one time and if the right neighbours are unique, \n",
    "                    then the node represented by this read will *not* get condensed and this node will add to the uncertainty\n",
    "                \"\"\"\n",
    "                if Repeat > 1 and len( set(RightNeighbors) ) != 1:\n",
    "                    \n",
    "                    #Checking for l_critical\n",
    "                    if Repeat > 2:\n",
    "                        # If repeat >=3, then the given read length is definitely less than L_critical.\n",
    "                        Is_less_than_critical_length = True\n",
    "                        Reason = \"Triple Repeats\"\n",
    "                    else:\n",
    "                        # We accumulate all the reads with length = 2 and check if they interleave.\n",
    "                        Position_of_repeat_less_than_2 += [ RepeatPositions ]\n",
    "                    \n",
    "                    # We count the number of times each neighbhour is repeats\n",
    "                    \"\"\"\n",
    "                        Example: If the repeat is \n",
    "                        AACA G\n",
    "                        AACA G\n",
    "                        AACA T\n",
    "                        AACA T\n",
    "                        AACA C\n",
    "                        AACA C\n",
    "                        AACA C\n",
    "                        AACA C\n",
    "                        AACA A\n",
    "                        AACA A\n",
    "                        AACA A\n",
    "                        \n",
    "                        Then the node [AACA] has 2 outgoing edges to ACA G, \n",
    "                        2 outgoing edges to ACA T, 4 outgoing edges to ACA C,\n",
    "                        3 outgoing edges to ACA A.\n",
    "                        \n",
    "                        Note, if AACA is always present as G(AACA), then the node GAAC\n",
    "                        will have all its outgoing edges to AACA and that wont be counted here because\n",
    "                        unique right neigbhour = 1\n",
    "                        \n",
    "                    \"\"\"\n",
    "                    Count_stats = Counter(RightNeighbors).values()\n",
    "                    \n",
    "                    Uncertainty = self.factlog(sum(Count_stats))\n",
    "                    for i in Count_stats:\n",
    "                        Uncertainty -= self.factlog(i)\n",
    "                    Gap = math.log(sum(Count_stats)/min(Count_stats),2)\n",
    "\n",
    "                    UpperboundUncertainty += [ Uncertainty ]\n",
    "                    UncertaintyGap += [ Gap ]\n",
    "                \"\"\"\n",
    "                    Re-initialization.\n",
    "                \"\"\"\n",
    "                LeftNeighbhors  =   read[0]\n",
    "                RightNeighbors  =   read[2]\n",
    "                RepeatPositions = [ read[3] ]\n",
    "\n",
    "                Repeat = 1\n",
    "            \n",
    "                CurrentString = read[1]\n",
    "        \n",
    "        if sorted(Position_of_repeat_less_than_2, key = itemgetter(0)) != sorted(Position_of_repeat_less_than_2, key = itemgetter(1)):\n",
    "           # print( Position_of_repeat_less_than_2, sorted(Position_of_repeat_less_than_2))\n",
    "            Reason = \"Interleaved Repeats\"\n",
    "            Is_less_than_critical_length = True\n",
    "            \n",
    "        Summary = [  self.ReadLength_Considered, sum(UpperboundUncertainty), sum(UpperboundUncertainty) - sum(UncertaintyGap) ,Is_less_than_critical_length, str(datetime.datetime.now()) ]\n",
    "        #print( \"ReadLength :\", self.ReadLength_Considered, \"Upperbound\", sum(UpperboundUncertainty), \"Lower Bound\", sum(UpperboundUncertainty) - sum(UncertaintyGap))\n",
    "        if False:\n",
    "            print(\"Length of the DNA is\", len(self.DNA_current))\n",
    "            print(\"Number of reads repeating of length\", self.ReadLength_Considered,\" is\", len(UpperboundUncertainty))\n",
    "            print(\"Time\", datetime.datetime.now())\n",
    "        #print(\"Uncertainity\", sum(UpperboundUncertainty), \"Is less than critical length\", Is_less_than_critical_length, \"reason\", Reason)\n",
    "        return(Summary)\n",
    "    \n",
    "\n",
    "filename = \"RhodobacterSphaeroides.fasta\"\n",
    "filename = \"Buchnera_aphidicola.fasta\"\n",
    "#filename = \"StaphylococcusAureus.fasta\"\n",
    "Gene = Genome(filename, 1)\n",
    "Gene.getUncertainty()\n",
    "#Gene.getReadLengthGraph()\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
