{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Length of the DNA is', 3188609)\n",
      "('Length', 2, 'Answer', [2, 18, 3112841, True, ['TG']])\n",
      "('Length', 3, 'Answer', [3, 64, 3178343, True, ['TTG']])\n",
      "('Length', 4, 'Answer', [4, 256, 3186729, True, ['TTTG']])\n",
      "('Length', 5, 'Answer', [5, 1025, 3188119, True, ['TTTTG']])\n",
      "('Length', 6, 'Answer', [6, 4098, 3188386, True, ['TTTTTG']])\n",
      "('Length', 7, 'Answer', [7, 16035, 3187797, True, ['TTTTTTG']])\n",
      "('Length', 8, 'Answer', [8, 54258, 3175659, True, ['TTTTTTTG']])\n",
      "('Length', 9, 'Answer', [9, 140800, 3099078, True, ['TTTTTTTGG']])\n",
      "('Length', 10, 'Answer', [10, 272240, 2836469, True, ['TTTTTTTGCC']])\n",
      "('Length', 11, 'Answer', [11, 380196, 2284678, True, ['TTTTTGGTACC']])\n",
      "('Length', 12, 'Answer', [12, 375917, 1509563, True, ['TTTTTGGTACCG']])\n",
      "('Length', 13, 'Answer', [13, 261820, 798368, True, ['TTTTTGGTACCGC']])\n",
      "('Length', 14, 'Answer', [14, 140810, 355504, True, ['TTTTTGGTACCGCC']])\n",
      "('Length', 15, 'Answer', [15, 57754, 130651, True, ['TTTTTGGTACCGCCA']])\n",
      "('Length', 16, 'Answer', [16, 23600, 50487, True, ['TTTTTGGTACCGCCAT']])\n",
      "('Length', 17, 'Answer', [17, 8890, 18565, True, ['TTTTTGGTACCGCCATT']])\n",
      "('Length', 18, 'Answer', [18, 3110, 6403, True, ['TTTTTGGTACCGCCATTC']])\n",
      "('Length', 19, 'Answer', [19, 1233, 2575, True, ['TTTTTGGTACCGCCATTCG']])\n",
      "('Length', 20, 'Answer', [20, 509, 1093, True, ['TTTTTGGTACCGCCATTCGT']])\n",
      "('Length', 21, 'Answer', [21, 183, 413, True, ['TTTTTGGTACCGCCATTCGTT']])\n",
      "('Length', 22, 'Answer', [22, 121, 301, True, ['TTTTTGGTACCGCCATTCGTTG']])\n",
      "('Length', 23, 'Answer', [23, 73, 187, True, ['TTTTTGGTACCGCCATTCGTTGG']])\n",
      "('Length', 24, 'Answer', [24, 42, 107, True, ['TTTTTGGTACCGCCATTCGTTGGT']])\n",
      "('Length', 25, 'Answer', [25, 34, 83, True, ['TTTTTGGTACCGCCATTCGTTGGTT']])\n",
      "('Length', 26, 'Answer', [26, 45, 140, True, ['TTTTTGGTACCGCCATTCGTTGGTTC']])\n",
      "('Length', 27, 'Answer', [27, 21, 54, True, ['TTTTTGGTACCGCCATTCGTTGGTTCG']])\n",
      "('Length', 28, 'Answer', [28, 26, 55, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGA']])\n",
      "('Length', 29, 'Answer', [29, 44, 116, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAA']])\n",
      "('Length', 30, 'Answer', [30, 10, 33, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAAT']])\n",
      "('Length', 31, 'Answer', [31, 10, 22, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATC']])\n",
      "('Length', 32, 'Answer', [32, 20, 67, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCC']])\n",
      "('Length', 33, 'Answer', [33, 11, 34, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCCA']])\n",
      "('Length', 34, 'Answer', [34, 8, 18, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCCAG']])\n",
      "('Length', 35, 'Answer', [35, 21, 44, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCCAGC']])\n",
      "('Length', 36, 'Answer', [36, 7, 22, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCCAGCC']])\n",
      "('Length', 37, 'Answer', [37, 12, 47, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCCAGCCG']])\n",
      "('Length', 38, 'Answer', [38, 18, 53, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCCAGCCGC']])\n",
      "('Length', 39, 'Answer', [39, 15, 37, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCCAGCCGCC']])\n",
      "('Length', 40, 'Answer', [40, 7, 25, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCCAGCCGCCC']])\n",
      "('Length', 41, 'Answer', [41, 15, 31, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCCAGCCGCCCC']])\n",
      "('Length', 42, 'Answer', [42, 2, 5, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCCAGCCGCCCCA']])\n",
      "('Length', 43, 'Answer', [43, 5, 16, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCCAGCCGCCCCAG']])\n",
      "('Length', 44, 'Answer', [44, 11, 47, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCCAGCCGCCCCAGC']])\n",
      "('Length', 45, 'Answer', [45, 4, 17, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCCAGCCGCCCCAGCC']])\n",
      "('Length', 46, 'Answer', [46, 2, 10, True, ['TTTTTGGTACCGCCATTCGTTGGTTCGAATCCAGCCGCCCCAGCCA']])\n",
      "('Length', 47, 'Answer', [47, 6, 20, True, ['TTTTCCGCTATGCCTCGCCTGACCAACCATTCGGGAGCCTCTCCATG']])\n",
      "('Length', 48, 'Answer', [48, 4, 8, True, ['TTTTCCGCTATGCCTCGCCTGACCAACCATTCGGGAGCCTCTCCATGT']])\n"
     ]
    }
   ],
   "source": [
    "# %load Genome.py\n",
    "# %load Genome.py\n",
    "from Bio import SeqIO\n",
    "import sys, numpy, math, datetime\n",
    "import CommonFunctions\n",
    "from CommonFunctions import WriteArrayinFile\n",
    "from operator import itemgetter\n",
    "from collections import Counter\n",
    "from doctest import Example\n",
    "\n",
    "class Genome:\n",
    "    DNAList= []\n",
    "    DNA_current = []\n",
    "    ReadLength_Considered = 100\n",
    "    Filename    = \"\"\n",
    "    SideLengths = 1\n",
    "    Gap         = 1\n",
    "    \"\"\"\n",
    "        Basic Initialization\n",
    "    \"\"\"\n",
    "    def __init__(self, Filename, Gap):\n",
    "        self.Gap = Gap\n",
    "        self.Filename = Filename\n",
    "        Handle = open(Filename)\n",
    "        for seq_record in SeqIO.parse(Handle, \"fasta\"):\n",
    "            self.DNAList.append(str(seq_record.seq))\n",
    "        print( \"Length of the DNA is\", len(self.DNAList[0]))\n",
    "    \n",
    "    \"\"\"\n",
    "        Implements log(n!)\n",
    "    \"\"\"\n",
    "    def factlog(self,n):\n",
    "        if n<20:\n",
    "            return(math.log(math.factorial(n), 2))\n",
    "        else:\n",
    "            return( n*(math.log(n,2) - 1.44269) + 0.5*math.log(2*math.pi*n)) #math.log(math.e,2) = 1.4426950408889634\n",
    "    \n",
    "    def getReadLengthGraph(self):\n",
    "     #Stores final data\n",
    "        Summary = []\n",
    "        for length in range(1,400):\n",
    "            length += self.Gap\n",
    "            Answer = self.NumberofRepeatsofLengthL(length) \n",
    "            if False == Answer[3]:\n",
    "                break  \n",
    "            print(\"Length\", length, \"Answer\", Answer)\n",
    "            Summary += [ Answer ]\n",
    "        CommonFunctions.WriteArrayinFile(Summary, \"Read_lengths\"+self.Filename[:-6]+\".csv\")\n",
    "           \n",
    "    def NumberofRepeatsofLengthL(self, length):\n",
    "        self.DNA_current = self.DNAList[0]\n",
    "        #print(\"DNA Length\", len(self.DNA_current))\n",
    "        Reads = []\n",
    "        for position in range(self.SideLengths, len(self.DNA_current) - self.ReadLength_Considered - self.SideLengths ):\n",
    "            Reads +=[   \n",
    "                        [ \n",
    "                         self.DNA_current[position - 1: position] , #Left Neighbhour\n",
    "                         self.DNA_current[position: position + length] ,\n",
    "                         self.DNA_current[position + length : position + length + 1],#Right Neigbhour\n",
    "                         position\n",
    "                        ]\n",
    "                    ]\n",
    "        #print(\"Sorting Read Array\")\n",
    "        # Sort w.r.t to the reads. (aggrerate them)\n",
    "        Reads = sorted(Reads,key=itemgetter(1) )\n",
    "        No_of_maximal_reads = 0\n",
    "        No_of_occurences_of_maximal_reads = 0\n",
    "        #print(\"Counting Repeats\")\n",
    "        CurrentString = Reads[0][1]\n",
    "        Neighbhors = []\n",
    "        Repeat = 1\n",
    "        Continue = False\n",
    "        Example = [\"none\"]\n",
    "        Positions = []\n",
    "        for read in Reads[1:]:\n",
    "            if read[1] == CurrentString:\n",
    "                Repeat += 1\n",
    "                Neighbhors += [ (read[0], read[2]) ]\n",
    "                Positions += [ read[3] ]\n",
    "            else:\n",
    "                \"\"\"\n",
    "                   Count only when both right and left neigbhors are not the same\n",
    "                \"\"\"\n",
    "                if Repeat > 1:\n",
    "                    Continue = True\n",
    "                    Example = [ CurrentString]\n",
    "                    Is_Maximal = False\n",
    "                    Neighbhors = list( set(Neighbhors) )\n",
    "                    for i in range(1,len(Neighbhors)):\n",
    "                        for j in range(i):\n",
    "                            if( Neighbhors[i][0] != Neighbhors[j][0] and Neighbhors[i][1] != Neighbhors[j][1] ): #Implies that the given read is maximal\n",
    "                                Is_Maximal = True\n",
    "                                break\n",
    "                        else:\n",
    "                            continue\n",
    "                        break\n",
    "                    if Is_Maximal:\n",
    "                        No_of_maximal_reads += 1\n",
    "                        No_of_occurences_of_maximal_reads += Repeat\n",
    "                            \n",
    "                Repeat = 1        \n",
    "                CurrentString = read[1]\n",
    "                Neighbhors = [ (read[0], read[2]) ]\n",
    "                Positions = [ read[3] ]\n",
    "        return( [ length, No_of_maximal_reads, No_of_occurences_of_maximal_reads, Continue,Example ])\n",
    "\n",
    "    def getUncertainty(self):\n",
    "        #Stores final data\n",
    "        Summary = []\n",
    "        length = 10\n",
    "        while True:\n",
    "            length += self.Gap\n",
    "            Answer = self.RepeatsofLengthL(length)  \n",
    "            #print(\"Answer\", Answer)\n",
    "            Summary += [ Answer ]\n",
    "            if Answer[-4] < 1:\n",
    "                break\n",
    "        CommonFunctions.WriteArrayinFile(Summary, \"New_Summary_\"+self.Filename[:-6]+\".csv\")\n",
    "   \n",
    "    def RepeatsofLengthL(self, length = 100):\n",
    "        self.ReadLength_Considered = length\n",
    "        self.DNA_current = self.DNAList[0]\n",
    "        #print(\"DNA Length\", len(self.DNA_current))\n",
    "        Reads = []\n",
    "        #print(\"Read Length\", length)\n",
    "        for position in range(self.SideLengths, len(self.DNA_current) - self.ReadLength_Considered - self.SideLengths ):\n",
    "         #   if position % 1000000 == 0:\n",
    "         #       print(\"In position\", position, \"DNA left\", len(self.DNA_current) - position )           \n",
    "            Reads +=[   \n",
    "                        [ \n",
    "                         self.DNA_current[position - self.SideLengths:                             position] , #Left Neighbhour\n",
    "                         self.DNA_current[position                   :position + self.ReadLength_Considered] ,\n",
    "                         self.DNA_current[position + self.ReadLength_Considered : position + self.ReadLength_Considered + self.SideLengths], #Right Neigbhour\n",
    "                         position\n",
    "                        ]\n",
    "                    ]\n",
    "        #print(\"Sorting Read Array\")\n",
    "        # Sort w.r.t to the reads. (aggrerate them)\n",
    "        Reads = sorted(Reads,key=itemgetter(1) )\n",
    "        UpperboundUncertainty = []\n",
    "        UncertaintyGap = []\n",
    "        \n",
    "        Is_less_than_critical_length = False\n",
    "        Reason = \"No reason\"\n",
    "        Position_of_repeat_less_than_2 = []\n",
    "        \n",
    "        #print(\"Counting Repeats\")\n",
    "        CurrentString = Reads[0][1]\n",
    "        Repeat = 0\n",
    "        LeftNeighbhors = []\n",
    "        RightNeighbors = []\n",
    "        RepeatPositions = []\n",
    "        Repeat = 1\n",
    "\n",
    "        for read in Reads[1:]:\n",
    "            if read[1] == CurrentString:\n",
    "                Repeat += 1\n",
    "                LeftNeighbhors  +=   read[0]\n",
    "                RightNeighbors  +=   read[2]\n",
    "                RepeatPositions += [ read[3] ]\n",
    "            else:\n",
    "                \"\"\"\n",
    "                    If the read repeats more than one time and if the right neighbours are unique, \n",
    "                    then the node represented by this read will *not* get condensed and this node will add to the uncertainty\n",
    "                \"\"\"\n",
    "                if Repeat > 1 and len( set(RightNeighbors) ) != 1:\n",
    "                    \n",
    "                    #Checking for l_critical\n",
    "                    if Repeat > 2:\n",
    "                        # If repeat >=3, then the given read length is definitely less than L_critical.\n",
    "                        Is_less_than_critical_length = True\n",
    "                        Reason = \"Triple Repeats\"\n",
    "                    else:\n",
    "                        # We accumulate all the reads with length = 2 and check if they interleave.\n",
    "                        Position_of_repeat_less_than_2 += [ RepeatPositions + read[1] ]\n",
    "                    \n",
    "                    # We count the number of times each neighbhour is repeats\n",
    "                    \"\"\"\n",
    "                        Example: If the repeat is \n",
    "                        AACA G\n",
    "                        AACA G\n",
    "                        AACA T\n",
    "                        AACA T\n",
    "                        AACA C\n",
    "                        AACA C\n",
    "                        AACA C\n",
    "                        AACA C\n",
    "                        AACA A\n",
    "                        AACA A\n",
    "                        AACA A\n",
    "                        \n",
    "                        Then the node [AACA] has 2 outgoing edges to ACA G, \n",
    "                        2 outgoing edges to ACA T, 4 outgoing edges to ACA C,\n",
    "                        3 outgoing edges to ACA A.\n",
    "                        \n",
    "                        Note, if AACA is always present as G(AACA), then the node GAAC\n",
    "                        will have all its outgoing edges to AACA and that wont be counted here because\n",
    "                        unique right neigbhour = 1\n",
    "                        \n",
    "                    \"\"\"\n",
    "                    Count_stats = Counter(RightNeighbors).values()\n",
    "                    \n",
    "                    Uncertainty = self.factlog(sum(Count_stats))\n",
    "                    for i in Count_stats:\n",
    "                        Uncertainty -= self.factlog(i)\n",
    "                    Gap = math.log(sum(Count_stats)/min(Count_stats),2)\n",
    "                    print( Count_stats, sum(Count_stats), min(Count_stats), Gap, Uncertainty)\n",
    "\n",
    "                    UpperboundUncertainty += [ Uncertainty ]\n",
    "                    UncertaintyGap += [ Gap ]\n",
    "                \"\"\"\n",
    "                    Re-initialization.\n",
    "                \"\"\"\n",
    "                LeftNeighbhors = []\n",
    "                RightNeighbors = []\n",
    "                RepeatPositions = []\n",
    "                Repeat = 1\n",
    "            \n",
    "                CurrentString = read[1]\n",
    "        \n",
    "        if sorted(Position_of_repeat_less_than_2, key = itemgetter(0)) != sorted(Position_of_repeat_less_than_2, key = itemgetter(1)):\n",
    "           # print( Position_of_repeat_less_than_2, sorted(Position_of_repeat_less_than_2))\n",
    "            Reason = \"Interleaved Repeats\"\n",
    "            Is_less_than_critical_length = True\n",
    "            \n",
    "        Summary = [ self.Filename, len(self.DNA_current), self.ReadLength_Considered, len(UpperboundUncertainty), sum(UpperboundUncertainty), sum(UpperboundUncertainty) - sum(UncertaintyGap) ,Is_less_than_critical_length, str(datetime.datetime.now()) ]\n",
    "        print( \"ReadLength :\", self.ReadLength_Considered, \"Upperbound\", sum(UpperboundUncertainty), \"Lower Bound\", sum(UpperboundUncertainty) - sum(UncertaintyGap))\n",
    "        if False:\n",
    "            print(\"Length of the DNA is\", len(self.DNA_current))\n",
    "            print(\"Number of reads repeating of length\", self.ReadLength_Considered,\" is\", len(UpperboundUncertainty))\n",
    "            print(\"Time\", datetime.datetime.now())\n",
    "        #print(\"Uncertainity\", sum(UpperboundUncertainty), \"Is less than critical length\", Is_less_than_critical_length, \"reason\", Reason)\n",
    "        return(Summary)\n",
    "\n",
    "\n",
    "filename = \"RhodobacterSphaeroides.fasta\"\n",
    "#filename = \"Buchnera_aphidicola.fasta\"\n",
    "#filename = \"StaphylococcusAureus.fasta\"\n",
    "Gene = Genome(filename, 1)\n",
    "# Gene.getUncertainty()\n",
    "Gene.getReadLengthGraph()\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
